### DISCUSSION TITLE
2.2.0 "Win7 style title bar button" description

### DISCUSSION THREAD

**Discussion by Maplespe** [Maplespe/DWMBlurGlass#282] [General]
In 2.2.0 we improved the Win7 style title bar button, now it should be consistent with Win7.
![O)NDDOF`6 4X0L0J)6GCET0](https://github.com/Maplespe/DWMBlurGlass/assets/43676879/bdfe09ea-66c3-4d97-adc9-4b943c88afdb)

Also added a config.ini setting
``` ini
[config]
titlebtnOffsetX=-1
```
reserved for those who need to change the offset when the automatically calculated value is incorrect, or when the offset needs to be changed.
The offset here is the margin (px) between the right of the close button and the right of the window.

**Comment by OldManAlpha** [comment]
Every day, we get a little bit closer to perfection.

**Comment by TheInsane101** [comment]
Amazing! Thank you so much for all your guys' hard work! 

It would be even more amazing if you could add Vista's caption sizes as an option, though that's not a priority!

**Comment by Maplespe** [comment]
I just implemented the button glow effect.
If you are interested you can download this test version and try it!
It only works if "`Restore Win7 style title bar button size`" is checked.
It has to work with DWMBlurGlass's title bar button modification, if you use other windhawk mods it may crash.
![test](https://github.com/Maplespe/DWMBlurGlass/assets/43676879/457416bf-48e2-4919-8184-5e343a852842)

I've only tested it on Windows 10 22H2, win11 may not work, I don't know.

[Release_x64_test.zip](https://github.com/Maplespe/DWMBlurGlass/files/15138245/Release_x64_test.zip)

**Reply by Maplespe** [reply]
I tested it and Windows 11 does work, but the glow is cropped by the rounded area.
![0WI L2Y4K0EAEG){RJ `@UW](https://github.com/Maplespe/DWMBlurGlass/assets/43676879/0632f228-b74f-42fa-b005-2e9e61bd8ad4)

**Reply by not-nullptr** [reply]
will test. quick question: does the titlebar button option also fix the window title margin?

**Reply by ghost** [reply]
Maple you god damn wizard, it works perfectly, even with Aero Window Manager on Windows 10 21H2

**Reply by Maplespe** [reply]
> will test. quick question: does the titlebar button option also fix the window title margin?

Not yet, but it's in the plans.

**Reply by kfh83** [reply]
On another note, do you plan on uploading these experimental features/bug fixes to the repo? It seems a lot of them just get buried in issues (and in this case, a discussion), and i think it would be useful for others to put this on the repo or maybe even a separate branch.

And also, this test version has automatic reinjection for anyone interested. This should also be in the base 2.2.0 release and most definitely in the repo.

**Reply by Maplespe** [reply]
I'll release a Release when all the work is done and upload the repository along with it.
This is for testing only, this alone should not be functional enough to release a new version.

**Reply by kfh83** [reply]
I'm aware it's not ready for release, but you could have a sort of "canary" branch with these experimental tweaks

**Reply by Maplespe** [reply]
Good idea.

**Reply by kfh83** [reply]
Thanks a lot.

**Reply by aubymori** [reply]
God bless you.

**Reply by TheInsane101** [reply]
This is AWESOME! To just jump in and add the glow effect when no one thought we'd ever see it again is such a pro move! (It works perfectly on my Windows 10 22H2 setup!) I had already decided that DWMBlurGlass was finally better than glass8 after 2.2, but this just blows my expectations even more! 

Could you add the ability for inactive caption buttons to retain their full opacity instead of fading? There's already a Windhawk mod for this, but it would be nice to have it included in DWMBG. Since such a mod already exists, I understand that it's a low priority.

Lastly, could you also add the ability to select reflection image files through the GUI? That's about all I can think of for making this more perfect than it already is! Amazing work, DWMBlurGlass team!

**Reply by Skyyblaze** [reply]
Thank you very much @Maplespe, I just tested the Glow build on Windows 11 23H2 and it works perfectly!

**Reply by OldManAlpha** [reply]
And here I was just thinking how it'd be nice if the buttons glowed, and then it just happens.

**Reply by TheInsane101** [reply]
Maplespe just popping in and being like "And oh yeah I also added the glow btw" without any warning is such a pro move

**Comment by kfh83** [comment]
https://github.com/Maplespe/DWMBlurGlass/assets/85626296/b3d785fe-c408-440f-8d90-5599c86df55c


You can get the glows to "work" without part of the Windows 7 caption shrink option in dwmbg that actually changes the sizes and offsets (this would mean you would have to use AWM or the windhawk mod for this), however...
![image](https://github.com/Maplespe/DWMBlurGlass/assets/85626296/68a51504-7027-4e03-bc32-2681c970f7b3)
Even if you do get the ability of changing your metrics back, the glows won't respond, and it seems like the win7 style titlebar button option only really accounts for DPI values and does some calculations off your chosen DPI value, not your system metrics.

**Reply by Maplespe** [reply]
Currently, you may need to reopen the window after changing the dpi to calculate the glow size correctly.
Also DWMBlurGlass calculates button sizes hardcoded to win7 sizes * dpi values.
DWMBlurGlass uses as much generic code as possible to avoid conflicts with those windhawk mods, but so I still don't recommend using it with similar mods.

The features themselves are all in one for DWMBlurGlass, I don't know why you would want to split them, it would be better to let them implement them instead of us trying to be compatible with them all the time.

**Reply by kfh83** [reply]
The second image is just a metrics change, not dpi BTW

**Reply by Maplespe** [reply]
So you want it to follow the system metrics ?

**Reply by aubymori** [reply]
Yes. It should be really easy with the `GetSystemMetricsForDpi` API and a bit of calculation from there.

**Reply by kfh83** [reply]
@Maplespe You could alternatively take this out of the Windows 7 captions option, and make it follow the actual size of the CButton (@wiktorwiktor12 should probably explain this better, im just relaying what i remember) since the captions option does change them (and so do literally any other means of reverting windows 10 buttons), might aswell just follow the size.

![image](https://github.com/Maplespe/DWMBlurGlass/assets/85626296/66388153-938a-4607-a1ae-68ea97b1cbc3)
![image](https://github.com/Maplespe/DWMBlurGlass/assets/85626296/7bd47660-31da-4b72-a62a-d362d8e212d6)

_(here's @wiktorwiktor12's modifications to CustomButton.cpp paired with @Dulappy's AWM. As you can see, it follows the size of the actual button and the system metrics.)_

**Reply by Maplespe** [reply]
It's a confusing behavior, why follow the system title bar height when you want to restore win7 button height? It destroys the original height.
If you change the title bar Height it is no longer win7 style.

And making an open source software compatible with a closed source internal testing software is a ridiculous thing to do.

**Reply by aubymori** [reply]
You could change it in Windows 7, you know... there's really no disadvantage to supporting this

**Reply by Maplespe** [reply]
For DWMBlurGlass's button recovery feature, the goal is to recover the win7 size and height of the button, if it follows the system metrics it's function is null and void So there's no reason to make it follow the system size.
For the button glow, this is just a test version, the actual version option will be separate and optional, it will follow the size of the CButton.

**Reply by aubymori** [reply]
The caption buttons on 7 changed size with the titlebar metric, but whatever.

**Reply by Maplespe** [reply]
You're right, but Windows 10, 11 also change button sizes following the system metric.
But this size is obviously too big Doesn't match win7, that's the whole point of this feature, if you let him keep it as it is, then it's not win7 style.

**Reply by aubymori** [reply]
This is how large titlebar metrics looked on Windows 7. This is objectively Win7 style. Whatever, this conversation isn't getting anywhere.

**Reply by Maplespe** [reply]
I think there is a problem with your formulation.
I guess what you are trying to achieve would be to replicate the metric calculation like Win7.
For Window10 the buttons should be scaled proportionally, not directly using the system metric.
Instead of using the system metric directly, it won't work, it'll get huge.

**Reply by aubymori** [reply]
Windows 10 button height is the sum of the titlebar height and the border width, so that the caption buttons take up the entire height of the titlebar. In Windows 7, the button height is just the titlebar height, not including the size of the borders. This is what makes them shorter in 7.

**Reply by Maplespe** [reply]
An example.
After reverting the buttons to win7 size and scaling the size according to DPI (this is the current behavior)
Then add scaling according to the system metric and scale again in equal proportions.
Not directly using the system metric as you said before, which is wrong.

**Reply by aubymori** [reply]
In any case, support for metrics is good to allow Vista users to use the DWMBG option. The default titlebar metric is 19 in Vista, and 21 in 7.

**Reply by aubymori** [reply]
> An example.
After reverting the buttons to win7 size and scaling the size according to DPI (this is the current behavior)
Then add scaling according to the system metric and scale again in equal proportions.
Not directly using the system metric as you said before, which is wrong.

I don't understand what you're trying to say.

**Reply by Maplespe** [reply]
I probably understand what you're saying, except that I'm basing my interpretation and understanding on a code level, so you may not understand what I'm talking about.

**Reply by aubymori** [reply]
Like I said earlier, you can get the titlebar metric with DPI support (`GetSystemMetricsForDpi`) so I don't really see a way that DPI and system metrics could cause a clash.

**Reply by Maplespe** [reply]
In short, replicating the title bar button size behavior, and metric calculation behavior in Win7.
I'm just saying that using modern Window's metric values directly won't work, and requires some additional calculations.
You don't have to understand that. Just make it the same as Win7 anyway, right?
I'll probably improve it later.

**Reply by kfh83** [reply]
I probably worded my initial request a bit too wrong, with metrics i meant the size of the actual button (which got understood in the end i guess).

**Comment by 7kt4** [comment]
For some reason, this version only works if I install it through the GUI. If I restart DWM, for example, it crashes and says that it has stopped injection.

**Reply by 7kt4** [reply]
That doesn't seem to happen now. But it has quite the memory leak...kind of ? 
Sometimes it is jumping between 400 and 800mb within only a few mins, and without opening many windows or anything

![image](https://github.com/Maplespe/DWMBlurGlass/assets/84914212/c306ffe2-199f-4d6c-88a6-ed56f5f3d76b)

And, after I uploaded this comment, it went back down to normal 100MB. 

So far it seems to always go back down once it reaches 800-900MB. So not a big deal if you have a lot of RAM, since there is a limit. And sometimes, it goes down without it becoming so high first.

**Reply by Maplespe** [reply]
I'm guessing you're using integrated graphics?

**Comment by 7kt4** [comment]
> I'm guessing you're using integrated graphics?

Yes...but I have watched DWM memory a lot in other versions, it's only now an issue with this buttons glow version. Both reaching the high levels, and the huge sudden falls/spikes. I restored one window, and it went up 50MB, then went back down about 10 seconds later.

**Reply by kfh83** [reply]
Caption buttons option is pretty experimental and causes a fair bit of crashes for me, would recommend to keep it off for now, after all, this _is_ a test version. Stay on regular 2.2.0.

**Comment by Ben2557** [comment]
> In 2.2.0 we improved the Win7 style title bar button, now it should be consistent with Win7. ![O)NDDOF`6 4X0L0J)6GCET0](https://private-user-images.githubusercontent.com/43676879/325996671-bdfe09ea-66c3-4d97-adc9-4b943c88afdb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE5MDI3MDUsIm5iZiI6MTcyMTkwMjQwNSwicGF0aCI6Ii80MzY3Njg3OS8zMjU5OTY2NzEtYmRmZTA5ZWEtNjZjMy00ZDk3LWFkYzktNGI5NDNjODhhZmRiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzI1VDEwMTMyNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU1OTYyZGRkN2YwN2ZlODY5NWNhMTE0ZmIzYmQ3ZjI2ZWE3ZTYzZmMyNTQ5MmZkNzY2YjE1YjJlMTBhZDJkZTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.7EKzV50YTJJXJEmrhe5v7go_MkMBkdxDbj5Tifvf2zQ)
> 
> Also added a config.ini setting
> 
> ```ini
> [config]
> titlebtnOffsetX=-1
> ```
> 
> reserved for those who need to change the offset when the automatically calculated value is incorrect, or when the offset needs to be changed. The offset here is the margin (px) between the right of the close button and the right of the window.

Can we manage to restore VISTA style ?
On Windows 11, it's impossible to use Windhawk mod like in Windows 10.

**Comment by kayodde** [comment]
How do you change it to look like windows 7 aero? I thought there was only an option to change the button size? In the OP image it has dwmblurglass with a windows 7 theme (unless this is another program)

**Reply by Sand216** [reply]
That's another program, probably either SecureUxTheme or UltraUxThemePatcher.

These programs allow you to use unsigned msstyles themes, as Microsoft locks them out by default.

I don't know the exact theme Maplespe is using here, but it is probably [Aero10](https://www.deviantart.com/vaporvance/art/Aero10-for-Windows-10-1903-22H2-909711949).

**Reply by Milicent-Heeler** [reply]
You can try out AeroLite with DWMBlurGlass. It looks similar to the windows 7 style.

**Comment by Waschbaermann420** [comment]
sup, so idk what im doing wrong, but i cant get the win7 style buttons. dont get me wrong, the standard one looks amazing but its not the win7 style yk? XD
![Screenshot (29)](https://github.com/user-attachments/assets/56a451dd-d090-4c85-b0f3-a9ae52e7f9d3)

**Reply by MrRazamataz** [reply]
You need to install a Windows 7 theme and have your system patched with UXThemePatcher. https://www.deviantart.com/vaporvance/art/Aero10-for-Windows-10-1903-22H2-909711949

**Reply by Undisputed00x** [reply]
> You need to install a Windows 7 theme and have your system patched with UXThemePatcher. https://www.deviantart.com/vaporvance/art/Aero10-for-Windows-10-1903-22H2-909711949

A better Windows 7 theme representation for Windows 11 would be https://github.com/SandTechStuff/Aero11

**Reply by Waschbaermann420** [reply]
> > You need to install a Windows 7 theme and have your system patched with UXThemePatcher. https://www.deviantart.com/vaporvance/art/Aero10-for-Windows-10-1903-22H2-909711949
> 
> A better Windows 7 theme representation for Windows 11 would be https://github.com/SandTechStuff/Aero11

Is it also compatible with Windows 10?
